Recursive list directory for FluentStorage.SFTP

beeradmoore opened the pull request:
### Fixes

Issue #55 

### Description

Not sure the preferred way to handle this. I created a local function so I could share the lock object rather than adding it as a class variable. If you are using .NET 6+ it allows the use of `Parallel.ForEachAsync`. Because of using `Parallel.ForEachAsync` I also added a lock for updating the main list. Unsure if there are complications of allowing this to query a server multiple times at once (also with no upper bound set).

I am also unsure if this should be littered with `.ConfigureAwait(false)` for all those tasks it is using.

I have not run any of the tests to confirm these changes work, but this is copy/pasted where I am using it for an internal tool I am testing out.

beeradmoore committed:
Recursive list directory for FluentStorage.SFTP

robinrodricks commented:
Can you extract the async method as a new method, because I don't like the pattern of methods being declared inside other methods.

robinrodricks commented on FluentStorage.SFTP/SshNetSftpBlobStorage.cs:
```diff
@@ -266,18 +266,43 @@ public SshNetSftpBlobStorage(string host, string username, params PrivateKeyFile
+			List<Blob> blobCollection = new List<Blob>();
+			object lockObject = new object();
+
+			async Task ListDirectoryAsync(string folderToList) {
+				IEnumerable<SftpFile> directoryContents = await client.ListDirectoryAsync(folderToList);
+				var tempBlobCollection = directoryContents
+					.Where(dc => (options.FilePrefix == null || dc.Name.StartsWith(options.FilePrefix))
+								 && (dc.IsDirectory || dc.IsRegularFile || dc.OwnerCanRead)
+								 && !cancellationToken.IsCancellationRequested
+								 && dc.Name != "."
+								 && dc.Name != "..")
+					.Take(options.MaxResults.Value)
+					.Select(ConvertSftpFileToBlob)
+					.Where(options.BrowseFilter).ToList();
+
+				lock (lockObject) {
+					blobCollection.AddRange(tempBlobCollection);
+				}
+
+				if (options.Recurse == true) {
+					IEnumerable<string> subFoldersToList = tempBlobCollection
+						.Where(x => x.IsFolder == true)
+						.Select(x => x.FullPath);
+#if NET6_0
+					await Parallel.ForEachAsync(subFoldersToList, async (subFolder, token) => {
+						await ListDirectoryAsync(subFolder);
+					});
+#else
+					foreach (string subFolder in subFoldersToList) { 
+						await ListDirectoryAsync(subFolder);
+					}
+#endif
+				}
+			}
```
extract this to a new method please

robinrodricks replied:
can you do this please?

beeradmoore replied:
Just finsihed for the day. I'll make this my first task tomorrow morning.

beeradmoore replied:
I moved the method out, bumped SSH.NET to 2023.0.1, and removed Renci.SshNet.Async. With that changed I had to change SftpClient to use ListDirectoryAsync which is an IAsyncEnumerable. I am unsure if I am using it correctly, or if it can be made more efficient. 

Compared to those other test results,

ParallelForEach = 3.0691516 sec
foreach = 17.7257746 sec
IAsyncEnumerable = 1.9855958 sec

robinrodricks commented on FluentStorage.SFTP/SshNetSftpBlobStorage.cs:
```diff
@@ -266,18 +266,43 @@ public SshNetSftpBlobStorage(string host, string username, params PrivateKeyFile
+						.Where(x => x.IsFolder == true)
+						.Select(x => x.FullPath);
+#if NET6_0
+					await Parallel.ForEachAsync(subFoldersToList, async (subFolder, token) => {
```
this will not help at all, there is only one SSH connection, so trying to do anything in parallel would not actually execute it in parallel against the server. convert this to a regular `for loop` please.

beeradmoore replied:
I removed this in my test project and the execution time increased dramatically, I had to double check I wasn't mistakingly connected to FTP which is much slower for us (we test between FTP and SFTP to the same server, same directory).

Our test directory it is 1877 files, across 192 folders.

ParallelForEach = 3.0691516 sec
foreach = 17.7257746 sec

For reference FTP is 43.6463812 sec. Turns out my initial thought of running the wrong test and connecting to FTP was off by a lot  😂

Even with one connection under the hood something else is happening there.

Would you like me to keep it in?

robinrodricks replied:
Ok, keep it in, I really dont understand how it works but maybe SSH.NET is using multiple threads in the background.

beeradmoore committed:
Moved ListDirectoryAsync to its own method

beeradmoore committed:
Removed Renci.SshNet.Async, bumped SSH.NET version

This changed involved re-writing SftpClient to use ListDirectoryAsync which is an IAsyncEnumerable.

beeradmoore committed:
Fixed naming convention of listDirectoryLockObject to match the rest of the class

Also marked as readonly to make static analysis happy.

robinrodricks commented:
Thanks a lot!
